### DISCUSSION TITLE
Allow VPC Peering

### DISCUSSION THREAD

**Discussion by dnk8n** [vercel/community#42] [Feature Question] [answered]
**Edit:** We now have a solution with [Vercel Secure Compute](https://vercel.com/docs/security/secure-compute).

---

Certain AWS features are protected with private only endpoints that can only be accessed when application is coming from same VPC. Are there plans to peer with Amazon VPC, to unlock these features?

For example, Amazon Serverless Postgress (Aurora) does not provide the ability to use a public IP. Connections have to come from within the same AWS VPC.

**Comment by bionicles** [comment]
I second this request — it’s incredibly risky for Zeit customers to secure production databases with only a password and no firewall. Even if the password remains secure, public databases **will** get DDOS’d on big sites, and there are many big sites using NextJS now, and many of these sites probably use Zeit, but without knowing IP ranges to allow, these big sites pretty much have to allow the whole interwebs. It’s well established now that DDOS can take down major swaths of the net, like what happened with Dyn DNS

At least give us a CLI command to lookup IP ranges to allow or let us grab those with Terraform outputs or something. You’re already managing global nameservers and could set up a few proxy load balancers to be allowed

**Comment by bionicles** [comment]
This is the number one reason I’d go to the trouble to use AWS over Zeit — Zeit has no easy way to connect the host servers to the database servers.
You want to use Zeit because it’s easy but this forces you to expose your databases. Not a good idea for the long run...

**Comment by dnk8n** [comment]
One idea I am going to test out is Flask, using SQL Alchemy, with Data API backend (I presume boto3 correctly setup in the Zeit Python runtime).

Found these libraries to do with accessing the RDS Data API (https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/data-api.html) via SqlAlchemy:

- https://github.com/koxudaxi/py-data-api
- https://github.com/rizidoro/camus
- https://github.com/chanzuckerberg/sqlalchemy-aurora-data-api

**Comment by phritz** [comment]
+1, this would be great. Seems like the only choices now are either to use the AWS RDS SDK in which case one can't use any standard ORM/db/etc packages that assume standard sql drivers, or have the endpoint public, which entails the DOS and cracking risks, and creates an attack surface where administration can happen from anywhere.

FWIW enabling lambda to access a vpc is a matter of passing the `vpc-config` flag to `lambda create-function`, see "Create the lambda function" at this link: https://docs.aws.amazon.com/lambda/latest/dg/services-rds-tutorial.html

**Comment by ghost** [comment]
I think a downside is that the start of a lambda is slower. Maybe this could be solved with provisioned concurrency?

**Reply by hervenivon** [reply]
The rollout of the improved VPC networking is finished since [august 2020](https://aws.amazon.com/blogs/compute/announcing-improved-vpc-networking-for-aws-lambda-functions/). Network interfaces are now created during the deployment phase, not at execution.

**Comment by wi-ski** [comment]
I'm from the future - would love to see this.

**Reply by chrisrickard** [reply]
Hold on, when exactly are you from in the future?
At least if we know it's not implemented by _then_ we can stop waiting.

**Comment by Aabhusan** [comment]
I am trying to connect the aurora-serverless database from the app running in vercel but the aurora-serverless database is only accessible within its vpc so I need to do vpc peering. can anyone help me with this??
Thanks in Advance

**Reply by Aabhusan** [reply]
@wi-ski @ivoreis @bionicles  @dnk8n  @ShanonJackson  @mjurincic  @phritz @r03 
please help if you guys have solved it

**Reply by phritz** [reply]
I tried and could not get aurora-serverless to work for the reason you state. I ended up getting provisioned aurora working with a static public IP, which obviously isn't awesome for a variety of reasons (e.g., cost and security). @rauchg solving this problem would enable services running on vercel that are backed by AWS cloud storage services to be faster, safer, and easier to use.

**Reply by phritz** [reply]
@elsigh ^

**Reply by ShanonJackson** [reply]
If you've got an experienced team and want VPC; this is your only option other than enterprise tier on Vercel. https://registry.terraform.io/modules/dealmore/next-js/aws/latest

This terraform mimics vercel's hosting platform incredibly closely.

**Comment by bionicles** [comment]
bump, still would use it, I truly love how easy it is to make apis with nextjs api routes and Vercel, but alas, this is a massive footgun

**Comment by leerob** [comment]
Hello! I wanted to give an update here. VPC peering is now on our roadmap. Feel free to [contact us](https://vercel.com/contact/sales) if this is something your team needs. Thank you 👍

**Reply by flybayer** [reply]
Will this be available on non-enterprise plans?

**Reply by leerob** [reply]
Hey @flybayer! We haven't figured that out yet. Will share more later 😄

**Reply by Phamiliarize** [reply]
Hopefully it could be really soon. Dying to use Vercel and Next.js at the company!

**Reply by bionicles** [reply]
I hope this is a non-enterprise feature because security in transit is critical to end-user privacy. "Enterprise" implies large scale -- not basic privacy.  

Small projects also have a duty to protect customer data.  If Vercel makes VPC peering accessible to all (obviously the account holder should pay the transfer cost), then more folks will be able to build apps with Vercel, and eventually scale those apps to the point where enterprise-level features become necessary

Self-service is awesome. We don't want to "contact sales" and sign up for a fancy enterprise edition just to provide commonsense features like "the database isn't exposed to the entire internet" -- Thanks for the good news, in any case!

**Reply by mobula9** [reply]
Hey @leerob, do you have any update about this feature? We are exciting to use it soon.

**Reply by GooBall** [reply]
This is the only thing preventing me from using vercel! It's an absolute must-have for production environments.

**Reply by flavianh** [reply]
Same here, I prototype on Vercel but would never use it as a production environment

**Reply by mobula9** [reply]
Hi @leerob I contacted the Vercel team on the form as you advice in this post but I never received any answer. (probably forgotten, you must be very busy)

(I precise I pay every month an entreprise plan)

**Reply by leerob** [reply]
Hello! Sorry for my slow response here. No updates yet, it is still on our roadmap. I will share back here when I have an update 👍

**Reply by ShanonJackson** [reply]
@leerob Any updates now? Its preventing us from migrating some projects onto vercel as we can't break our network boundries.

**Reply by mobula9** [reply]
Any update @leerob ? Thanks

**Reply by franco-roura** [reply]
Hi there 👋  Really excited about this feature, I'd love to move all of our AWS backend infrastructure to Vercel as soon as it's available.
@leerob do you think this is something possible for 2022 Q4? Or should we expect it to arrive somewhere in the future?

**Reply by DavidChouinard** [reply]
Is VPC peering available only for Enterprise plans?

**Comment by mobula9** [comment]
I would love to hear from Vercel on this. I will try to explain to you why this is so important to many of us.

It's essential to respect the confidentiality of our users for our business… but also more generally for ethical questions.

And being able to use our own VPC would allow us to use Vercel's superb functionalities for the entire stack, not just the frontend part but also the backend (through which our users' sensitive data passes).

Today, to respect our commitment to our users, we are obliged to maintain the deployment of the backend outside Vercel.

It's sad, because we have to do the work in duplicate and so, Vercel's interest is lost since we ultimately still have to manage part of the deployments ourselves.

This functionality would therefore open up the use of Vercel for all serious companies which really respect the confidentiality of their client.

This is the case with my company, and as long as Vercel does not allow this functionality, all the backend part will remain executed outside the framework of Vercel. And only the front end will go through Vercel (since the data does not go through the frontend code).

**Comment by mobula9** [comment]
It been now more than one year @leerob told this was on the roadmap. We are many to still wait this feature with impatience.. Any news @leerob ? 🤞

**Comment by jperelli** [comment]
Hi, any news about this? It's the only blocker for us to move our platform from AWS ECS to vercel

**Comment by pirtlj** [comment]
@leerob please give an update here,  this is also a big issue for us.

**Comment by darwin403** [comment]
Would love this as well! All our backend is on AWS private subnets which are ideally never to the public. It would be huge if our web application can talk to those services via VPC peeing!: )

**Comment by mobula9** [comment]
Hello @leerob we are still waiting for news about this. Please give an update we are waiting for do many time. I'm starting to think to leave nextjs just for this reason. Too bad.

**Comment by jason-nivelo** [comment]
Any updates

**Comment by leerob** [comment, marked as answer]
**Edit:** We now have a solution with [Vercel Secure Compute](https://vercel.com/docs/security/secure-compute).

**Reply by adithyakothapalli7** [reply]
Hi @leerob, our team is waiting for this. any update?

**Reply by swarnava** [reply]
Please [contact us](https://vercel.com/contact/sales) if this is something your team needs.

**Reply by franco-roura** [reply]
Dude, we'll finally be able to host SOC2, HIPAA, and PCI DSS compliant solutions on Vercel, this is **HUGE** 🎉  

Congrats to the team, you folks are killing it 🔥
